File: gdx2d-desktop/src/hevs/gdx2d/demos/tilemap/advanced/DemoTileAdvanced.java
package hevs.gdx2d.demos.tilemap.advanced;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.maps.tiled.*;
import com.badlogic.gdx.maps.tiled.renderers.OrthogonalTiledMapRenderer;
import com.badlogic.gdx.math.Vector2;

import hevs.gdx2d.lib.GdxGraphics;
import hevs.gdx2d.lib.PortableApplication;

import java.util.Map;
import java.util.TreeMap;

/**
 * Character walks on a small desert. (Tiled editor (http://www.mapeditor.org/)
 *
 * @author Alain Woeffray (woa)
 * @author Pierre-André Mudry (mui)
 * @author Marc Pignat (pim)
 */
public class DemoTileAdvanced extends PortableApplication {

    // key management
    private Map<Integer, Boolean> keyStatus = new TreeMap<Integer, Boolean>();

    // character
    private Hero hero;

    // tiles management
    private TiledMap tiledMap;
    private TiledMapRenderer tileMapRenderer;
    private TiledMapTileLayer tileLayer;

    @Override
    public void onInit() {

        // create hero
        hero = new Hero(10,20);

        // init keys status
        keyStatus.put(Input.Keys.UP, false);
        keyStatus.put(Input.Keys.DOWN, false);
        keyStatus.put(Input.Keys.LEFT, false);
        keyStatus.put(Input.Keys.RIGHT, false);

        // create map
        tiledMap = new TmxMapLoader().load("data/maps/desert.tmx");
        tileMapRenderer = new OrthogonalTiledMapRenderer(tiledMap);
        tileLayer = (TiledMapTileLayer)tiledMap.getLayers().get(0);
    }

    @Override
    public void onGraphicRender(GdxGraphics g) {
        g.clear();

        // Hero activity
        manageHero();
      
        // Camera follows the hero
        moveCamera(g);

        // Render the tilemap
        tileMapRenderer.setView(g.getCamera());
        tileMapRenderer.render();

        // Draw the hero
        hero.animate(Gdx.graphics.getDeltaTime());
        hero.draw(g);


        g.drawFPS();
        g.drawSchoolLogo();
    }

    /**
     * Move the camera to the hero's position
     * @param g Graphics object
     */
    private void moveCamera(GdxGraphics g){
        int dx = (int)(hero.getPosition().x - g.getCamera().position.x);
        int dy = (int)(hero.getPosition().y - g.getCamera().position.y);
        g.scroll(dx, dy);
    }

    /**
     * exemple : getTile(myPosition,0,1) get the tile over myPosition
     *
     * @param position The position on map (not on screen)
     * @param offsetX The number of cells at right of the given position.
     * @param offsetY The number of cells over the given position.
     * @return The tile around the given position | null
     */
    private TiledMapTile getTile(Vector2 position, int offsetX, int offsetY){
        try {
            int x = (int)(position.x / tileLayer.getTileWidth()) + offsetX;
            int y = (int)(position.y / tileLayer.getTileHeight()) + offsetY;

            return tileLayer.getCell(x, y).getTile();
        }catch( Exception e){

            return null;
        }
    }

    /**
     * Get the "walkable" property of the given tile.
     * @param tile The tile to know the property
     * @return true if the property is set to "true", false otherwise
     */
    private boolean isWalkable(TiledMapTile tile){
        if(tile == null)
            return false;

        Object test = tile.getProperties().get("walkable");

        return Boolean.parseBoolean(test.toString());
    }

    /**
     * Get the "speed" property of the given tile.
     * @param tile The tile to know the property
     * @return The value of the property
     */
    private float getSpeed(TiledMapTile tile){

        Object test = tile.getProperties().get("speed");

        return Float.parseFloat(test.toString());
    }

    /**
     * Manage the movements of the hero using the keyboard.
     */
    private void manageHero(){

    	// Do nothing if hero is already moving
        if(!hero.isMoving()){

        	
        	// Compute direction and next cell
            TiledMapTile nextCell = null;
            Hero.Direction goalDirection = Hero.Direction.NULL;

            if(keyStatus.get(Input.Keys.RIGHT)){
                goalDirection = Hero.Direction.RIGHT;
                nextCell = getTile(hero.getPosition(), 1, 0);
            }else if(keyStatus.get(Input.Keys.LEFT)){
                goalDirection = Hero.Direction.LEFT;
                nextCell = getTile(hero.getPosition(), -1, 0);
            }else if(keyStatus.get(Input.Keys.UP)){
                goalDirection = Hero.Direction.UP;
                nextCell = getTile(hero.getPosition(), 0, 1);
            }else if(keyStatus.get(Input.Keys.DOWN)){
                goalDirection = Hero.Direction.DOWN;
                nextCell = getTile(hero.getPosition(), 0, -1);
            }

            // Is the move valid ?
            if( isWalkable(nextCell) ) {
            	// Go
                hero.setSpeed(getSpeed(nextCell));
                hero.go(goalDirection);
            }else{
            	// Face the wall
                hero.turn(goalDirection);
            }
        }
    }

    // Manage keyboard events
    @Override
    public void onKeyUp(int keycode) {
        super.onKeyUp(keycode);

        keyStatus.put(keycode, false);
    }

    @Override
    public void onKeyDown(int keycode) {
        super.onKeyDown(keycode);

        keyStatus.put(keycode, true);
    }

    public static void main(String[] args) {
        new DemoTileAdvanced();
    }
}

File: gdx2d-demoDesktop/src/main/scala/ch/hevs/gdx2d/demos/tilemap/advanced/DemoTileAdvanced.scala
package ch.hevs.gdx2d.demos.tilemap.advanced

import scala.collection.mutable

import ch.hevs.gdx2d.desktop.DesktopApplication
import ch.hevs.gdx2d.lib.GdxGraphics
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.maps.tiled._
import com.badlogic.gdx.maps.tiled.renderers.OrthogonalTiledMapRenderer
import com.badlogic.gdx.math.Vector2

/**
 * Character walks on a small desert. (Tiled editor (http://www.mapeditor.org/)
 *
 * @author Alain Woeffray (woa)
 * @author Pierre-André Mudry (mui)
 * @author Marc Pignat (pim)
 * @version 2.0
 */
class DemoTileAdvanced extends DesktopApplication {

  private val keyStatus = new mutable.HashMap[Int, Boolean]()

  private var hero: Hero = _
  private var tiledMap: TiledMap = _
  private var tiledMapRenderer: TiledMapRenderer = _
  private var tiledLayer: TiledMapTileLayer = _
  private var zoomFactor: Float = 1f

  override def onInit(): Unit = {
    hero = new Hero(10, 20)
    zoomFactor = 1f

    keyStatus.put(Input.Keys.UP, false)
    keyStatus.put(Input.Keys.DOWN, false)
    keyStatus.put(Input.Keys.LEFT, false)
    keyStatus.put(Input.Keys.RIGHT, false)

    tiledMap = new TmxMapLoader().load("maps/desert.tmx")
    tiledMapRenderer = new OrthogonalTiledMapRenderer(tiledMap)
    tiledLayer = tiledMap.getLayers.get(0).asInstanceOf[TiledMapTileLayer]
  }

  override def onGraphicRender(g: GdxGraphics): Unit = {
    g.clear()

    manageHero()

    g.zoom(zoomFactor)
    g.moveCamera(hero.position.x, hero.position.y, tiledLayer.getWidth * tiledLayer.getTileWidth, tiledLayer.getHeight * tiledLayer.getTileHeight)

    tiledMapRenderer.setView(g.getCamera)
    tiledMapRenderer.render()

    hero.animate(Gdx.graphics.getDeltaTime.toDouble)
    hero.draw(g)

    g.drawFPS()
    g.drawSchoolLogo()
  }

  private def getTile(position: Vector2, offsetX: Int, offsetY: Int): TiledMapTile = {
    try {
      val x = (position.x / tiledLayer.getTileWidth).toInt + offsetX
      val y = (position.y / tiledLayer.getTileHeight).toInt + offsetY
      tiledLayer.getCell(x, y).getTile
    } catch {
      case _: Exception => null
    }
  }

  private def isWalkable(tile: TiledMapTile): Boolean = {
    if (tile == null) return false
    val test = tile.getProperties.get("walkable")
    if (test == null) return false
    java.lang.Boolean.parseBoolean(test.toString)
  }

  private def getSpeed(tile: TiledMapTile): Float = {
    val test = tile.getProperties.get("speed")
    if (test == null) return 1.0f
    test.toString.toFloat
  }

  private def manageHero(): Unit = {
    if (!hero.isMoving) {
      var nextCell: TiledMapTile = null
      var goalDirection: Hero.Direction.Value = Hero.Direction.NULL

      if (keyStatus.getOrElse(Input.Keys.RIGHT, false)) {
        goalDirection = Hero.Direction.RIGHT
        nextCell = getTile(hero.position, 1, 0)
      } else if (keyStatus.getOrElse(Input.Keys.LEFT, false)) {
        goalDirection = Hero.Direction.LEFT
        nextCell = getTile(hero.position, -1, 0)
      } else if (keyStatus.getOrElse(Input.Keys.UP, false)) {
        goalDirection = Hero.Direction.UP
        nextCell = getTile(hero.position, 0, 1)
      } else if (keyStatus.getOrElse(Input.Keys.DOWN, false)) {
        goalDirection = Hero.Direction.DOWN
        nextCell = getTile(hero.position, 0, -1)
      }

      if (goalDirection != Hero.Direction.NULL) {
        if (isWalkable(nextCell)) {
          hero.speed = getSpeed(nextCell)
          hero.go(goalDirection)
        } else {
          hero.turn(goalDirection)
        }
      }
    }
  }

  override def onKeyUp(keycode: Int): Unit = {
    super.onKeyUp(keycode)
    keyStatus.put(keycode, false)
  }

  override def onKeyDown(keycode: Int): Unit = {
    super.onKeyDown(keycode)
    keycode match {
      case Input.Keys.Z =>
        if (zoomFactor == 1.0f) zoomFactor = .5f
        else if (zoomFactor == .5f) zoomFactor = 2f
        else zoomFactor = 1f
      case _ =>
        keyStatus.put(keycode, true)
    }
  }

  override def onDispose(): Unit = {
    super.onDispose()
    tiledMap.dispose()
  }
}

object DemoTileAdvanced {
  def main(args: Array[String]): Unit = {
    new DemoTileAdvanced().launch()
  }
}
